apisix/plugins/api-breaker.lua

```diff
@@ -0,0 +1,212 @@
+                    type = "array",
+                    minItems = 1,
+                    items = {
+                        type = "integer",
+                        minimum = 500,
+                        maximum = 599,
+                    },
+                    uniqueItems = true,
+                    default = {500}
+                },
+                failures = {
+                    type = "integer",
+                    minimum = 1,
+                    default = 1,
+                }
+            }
+        },
+        healthy = {
+            type = "object",
+            properties = {
+                http_statuses = {
+                    type = "array",
+                    minItems = 1,
+                    items = {
+                        type = "integer",
+                        minimum = 200,
+                        maximum = 499,
+                    },
+                    uniqueItems = true,
+                    default = {200, 206}
+                },
+                successes = {
+                    type = "integer",
+                    minimum = 1,
+                    default = 1,
+                }
+            }
+        }
+    },
+    required = {"unhealthy_response_code", "unhealthy", "healthy"},
+}
+
+
+local function is_unhealthy(unhealthy_status, upstream_statu)
+    for _, unhealthy in ipairs(unhealthy_status) do
+        if unhealthy == upstream_statu then
+            return true
+        end
+    end
+
+    return false
+end
+
+
+local function is_healthy(healthy_status, upstream_statu)
+    for _, healthy in ipairs(healthy_status) do
+        if healthy == upstream_statu then
+            return true
+        end
+    end
+
+    return false
+end
+
+
+local function healthy_cache_key(ctx)
+    return "healthy-" .. core.request.get_host(ctx) .. ctx.var.uri
+end
+
+
+local function unhealthy_cache_key(ctx)
+    return "unhealthy-" .. core.request.get_host(ctx) .. ctx.var.uri
+end
+
+
+local function unhealthy_lastime_cache_key(ctx)
+    return "unhealthy-lastime" .. core.request.get_host(ctx) .. ctx.var.uri
+end
+
+
+local _M = {
+    version = 0.1,
+    name = plugin_name,
+    priority = 1005,
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return false, err
+    end
+
+    return true
+end
+
+
+function _M.access(conf, ctx)
+    local unhealthy_val, err = shared_buffer:get(unhealthy_cache_key(ctx))
+    if err then
+        core.log.error("ngx.shared get error", err)
+    end
+
+    local unhealthy_lastime, err = shared_buffer:get(unhealthy_lastime_cache_key(ctx))
+    if err then
+        core.log.error("ngx.shared get error", err)
+    end
+
+    if unhealthy_val and unhealthy_lastime then
+        local ride = math.ceil(unhealthy_val / conf.unhealthy.failures)
+        if ride < 1 then
+            ride = 1
+        end
+
+        if unhealthy_lastime + 2^ride >= ngx.time() then
```

Comment:
Hard code? And not has the max limit?

Reply:
What is your opinion about "hard code"? If there is a maximum limit, what is the logic when the maximum limit is triggered?

Reply:
@liuhengloveyou we need to write this into the doc, the user needs to know it.

Reply:
OK。This algorithm, I describe in the documentation。
With regard to user configurability, I think it could be TODO first. wait until there is user demand before adding it.

Reply:
I've updated the doc. Take a look.